src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -215,10 +236,25 @@ SendCommandToWorkersParams(TargetWorkerSet targetWorkerSet, char *command,
 												parameterTypes, parameterValues);
 		if (querySent == 0)
 		{
-			ReportConnectionError(connection, ERROR);
+			if (targetWorkerSet == WORKERS_WITH_METADATA)
+			{
+				ReportConnectionError(connection, WARNING);
+				MarkNodeMetadataSynced(connection->hostname, connection->port, false);
```

Comment:
Maybe we can leave these functions as they are and just have some alternative implementation for master_update_node?

Most of the time, having atomic metadata updates (i.e. erroring out on failure) seems preferable to setting metadatasynced to false. We also need to preserve a working, fault-tolerant MX implementation (MX = clients connect directly to workers). master_update_node is different because we need to make progress while a node is down for the system to recover. We'll need to figure out a separate solution for clients connecting to stale nodes (e.g. take node out of load balancer until synced), but at least that will be limited to failover scenarios.

Reply:
Makes sense. Today I couldn't get this done (maintenance daemon related issues took time). Our changes to `SendCommandToWorkers()` weren't even correct, since on error it just left the worker connection in error state. I'll add a `SendOptionalCommandToWorkers()` which uses savepoints to rollback errors.